File: src/main/java/org/railwaystations/rsapi/adapter/web/controller/InboxController.java
package org.railwaystations.rsapi.adapter.web.controller;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.commons.io.IOUtils;
import org.apache.commons.io.output.NullOutputStream;
import org.apache.commons.lang3.StringUtils;
import org.railwaystations.rsapi.app.auth.AuthUser;
import org.railwaystations.rsapi.app.auth.RSAuthenticationProvider;
import org.railwaystations.rsapi.app.auth.RSUserDetailsService;
import org.railwaystations.rsapi.core.model.*;
import org.railwaystations.rsapi.core.services.InboxService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.web.servlet.ModelAndView;

import javax.servlet.http.HttpServletRequest;
import javax.validation.constraints.NotNull;
import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.List;

@RestController
public class InboxController {

    private static final Logger LOG = LoggerFactory.getLogger(InboxController.class);

    private static final ObjectMapper MAPPER = new ObjectMapper();
    public static final String EMAIL = "email";
    public static final String UPLOAD_TOKEN = "uploadToken";
    public static final String STATION_ID = "stationId";
    public static final String COUNTRY_CODE = "countryCode";
    public static final String STATION_TITLE = "stationTitle";
    public static final String LATITUDE = "latitude";
    public static final String LONGITUDE = "longitude";
    public static final String COMMENT = "comment";
    public static final String ACTIVE = "active";
    public static final String FILE = "file";

    private final InboxService inboxService;
    private final RSAuthenticationProvider authenticator;
    private final RSUserDetailsService userDetailsService;

    public InboxController(final InboxService inboxService, final RSAuthenticationProvider authenticator,
                           final RSUserDetailsService userDetailsService) {
        this.inboxService = inboxService;
        this.authenticator = authenticator;
        this.userDetailsService = userDetailsService;
    }

    /**
     * Not part of the "official" API.
     * Supports upload of photos via the website.
     */
    @PostMapping(consumes = MediaType.MULTIPART_FORM_DATA_VALUE, value = "/photoUpload", produces = MediaType.TEXT_HTML_VALUE)
    @ResponseBody
    public ModelAndView photoUploadIframe(@RequestHeader(value = HttpHeaders.USER_AGENT, required = false) final String userAgent,
                                          @RequestParam(EMAIL) final String email,
                                          @RequestParam(UPLOAD_TOKEN) final String uploadToken,
                                          @RequestParam(value = STATION_ID, required = false) final String stationId,
                                          @RequestParam(value = COUNTRY_CODE, required = false) final String countryCode,
                                          @RequestParam(value = STATION_TITLE, required = false) final String stationTitle,
                                          @RequestParam(value = LATITUDE, required = false) final Double latitude,
                                          @RequestParam(value = LONGITUDE, required = false) final Double longitude,
                                          @RequestParam(value = COMMENT, required = false) final String comment,
                                          @RequestParam(value = ACTIVE, required = false) final Boolean active,
                                          @RequestParam(value = FILE) final MultipartFile file,
                                          @RequestHeader(value = HttpHeaders.REFERER) final String referer) throws JsonProcessingException {
        LOG.info("MultipartFormData: email={}, station={}, country={}, file={}", email, stationId, countryCode, file.getName());
        final URI refererUri = URI.create(referer);

        try {
            final Authentication authentication = authenticator.authenticate(new UsernamePasswordAuthenticationToken(email, uploadToken));
            if (authentication == null || !authentication.isAuthenticated()) {
                return createIFrameAnswer(consumeBodyAndReturn(file.getInputStream(), new InboxResponse(InboxResponse.InboxResponseState.UNAUTHORIZED)), refererUri);
            }

            final InboxResponse response = uploadPhoto(userAgent, file.getInputStream(), StringUtils.trimToNull(stationId),
                    StringUtils.trimToNull(countryCode), file.getContentType(), stationTitle, latitude, longitude, comment, active, userDetailsService.loadUserByUsername(email));
            return createIFrameAnswer(response, refererUri);
        } catch (final Exception e) {
            LOG.error("FormUpload error", e);
            return createIFrameAnswer(new InboxResponse(InboxResponse.InboxResponseState.ERROR), refererUri);
        }
    }

    @PostMapping(consumes = MediaType.ALL_VALUE ,produces = MediaType.APPLICATION_JSON_VALUE, value = "/photoUpload")
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<InboxResponse> photoUpload(final HttpServletRequest request,
                                                     @RequestHeader(value = HttpHeaders.USER_AGENT, required = false) final String userAgent,
                                                     @RequestHeader(value = "Station-Id", required = false) final String stationId,
                                                     @RequestHeader(value = "Country", required = false) final String country,
                                                     @RequestHeader(value = HttpHeaders.CONTENT_TYPE) final String contentType,
                                                     @RequestHeader(value = "Station-Title", required = false) final String encStationTitle,
                                                     @RequestHeader(value = "Latitude", required = false) final Double latitude,
                                                     @RequestHeader(value = "Longitude", required = false) final Double longitude,
                                                     @RequestHeader(value = "Comment", required = false) final String encComment,
                                                     @RequestHeader(value = "Active", required = false) final Boolean active,
                                                     @AuthenticationPrincipal final AuthUser user) throws IOException {
        final String stationTitle = encStationTitle != null ? URLDecoder.decode(encStationTitle, StandardCharsets.UTF_8) : null;
        final String comment = encComment != null ? URLDecoder.decode(encComment, StandardCharsets.UTF_8) : null;
        final InboxResponse inboxResponse = uploadPhoto(userAgent, request.getInputStream(), StringUtils.trimToNull(stationId),
                StringUtils.trimToNull(country), contentType, stationTitle, latitude, longitude, comment, active, user);
        return new ResponseEntity<>(inboxResponse, map(inboxResponse.getState()));
    }

    @PostMapping(consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE, value = "/reportProblem")
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<InboxResponse> reportProblem(@RequestHeader(HttpHeaders.USER_AGENT) final String userAgent,
                                       @RequestBody @NotNull() final ProblemReport problemReport,
                                       @AuthenticationPrincipal final AuthUser user) {
        final InboxResponse inboxResponse = inboxService.reportProblem(problemReport, user.getUser(), userAgent);
        return new ResponseEntity<>(inboxResponse, map(inboxResponse.getState()));
    }

    @GetMapping(produces = MediaType.APPLICATION_JSON_VALUE, value = "/publicInbox")
    public List<PublicInboxEntry> publicInbox() {
        return inboxService.publicInbox();
    }

    @PostMapping(consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE, value = "/userInbox")
    @PreAuthorize("isAuthenticated()")
    public List<InboxStateQuery> userInbox(@AuthenticationPrincipal final AuthUser user, @RequestBody @NotNull final List<InboxStateQuery> queries) {
        return inboxService.userInbox(user.getUser(), queries);
    }

    @GetMapping(produces = MediaType.APPLICATION_JSON_VALUE, value = "/adminInbox")
    @PreAuthorize("hasRole('ADMIN')")
    public List<InboxEntry> adminInbox(@AuthenticationPrincipal final AuthUser user) {
        return inboxService.listAdminInbox(user.getUser());
    }

    @PostMapping(consumes = MediaType.APPLICATION_JSON_VALUE, value = "/adminInbox", produces = MediaType.APPLICATION_JSON_VALUE)
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<AdminInboxCommandResponse> adminInbox(@AuthenticationPrincipal final AuthUser user, @RequestBody final InboxEntry command) {
        try {
            inboxService.processAdminInboxCommand(user.getUser(), command);
        } catch (final IllegalArgumentException e) {
            LOG.warn("adminInbox command {} failed", command, e);
            return new ResponseEntity<>(new AdminInboxCommandResponse(HttpStatus.BAD_REQUEST.value(), e.getMessage()), HttpStatus.BAD_REQUEST);
        }

        return new ResponseEntity<>(new AdminInboxCommandResponse(HttpStatus.OK.value(), "ok"), HttpStatus.OK);
    }

    @GetMapping(produces = MediaType.APPLICATION_JSON_VALUE, value = "/adminInboxCount")
    @PreAuthorize("hasRole('ADMIN')")
    public InboxCountResponse adminInboxCount(@AuthenticationPrincipal final AuthUser user) {
        return new InboxCountResponse(inboxService.countPendingInboxEntries());
    }

    @GetMapping(produces = MediaType.APPLICATION_JSON_VALUE, value = "/nextZ")
    public NextZResponse nextZ() {
        return new NextZResponse(inboxService.getNextZ());
    }

    private InboxResponse uploadPhoto(final String userAgent, final InputStream body, final String stationId,
                                      final String country, final String contentType, final String stationTitle,
                                      final Double latitude, final Double longitude, final String comment,
                                      final Boolean active, final AuthUser user) {
        final InboxResponse inboxResponse = inboxService.uploadPhoto(userAgent, body, stationId, country, contentType, stationTitle,
                latitude, longitude, comment, active, user.getUser());
        return consumeBodyAndReturn(body, inboxResponse);
    }

    private HttpStatus map(final InboxResponse.InboxResponseState state) {
        return switch (state) {
            case REVIEW -> HttpStatus.ACCEPTED;
            case LAT_LON_OUT_OF_RANGE, NOT_ENOUGH_DATA, UNSUPPORTED_CONTENT_TYPE -> HttpStatus.BAD_REQUEST;
            case PHOTO_TOO_LARGE -> HttpStatus.PAYLOAD_TOO_LARGE;
            case CONFLICT -> HttpStatus.CONFLICT;
            case UNAUTHORIZED -> HttpStatus.UNAUTHORIZED;
            case ERROR -> HttpStatus.INTERNAL_SERVER_ERROR;
        };
    }

    private ModelAndView createIFrameAnswer(final InboxResponse response, final URI referer) throws JsonProcessingException {
        final ModelAndView modelAndView = new ModelAndView();
        modelAndView.setViewName("iframe");
        modelAndView.getModel().put("response", MAPPER.writeValueAsString(response));
        modelAndView.getModel().put("referer", referer);
        return modelAndView;
    }

    private InboxResponse consumeBodyAndReturn(final InputStream body, final InboxResponse response) {
        if (body != null) {
            try {
                IOUtils.copy(body, NullOutputStream.NULL_OUTPUT_STREAM);
            } catch (final IOException e) {
                LOG.warn("Unable to consume body", e);
            }
        }
        return response;
    }

    private record InboxCountResponse(int pendingInboxEntries) {

        public int getPendingInboxEntries() {
            return pendingInboxEntries;
        }
    }

    private record NextZResponse(String nextZ) {

        public String getNextZ() {
            return nextZ;
        }

    }

    private record AdminInboxCommandResponse(int status, String message) {

        public int getStatus() {
            return status;
        }

        public String getMessage() {
            return message;
        }
    }

}

File: src/main/kotlin/org/railwaystations/rsapi/adapter/web/controller/InboxController.kt
package org.railwaystations.rsapi.adapter.web.controller

import jakarta.validation.Valid
import org.railwaystations.rsapi.adapter.web.InboxResponseMapper
import org.railwaystations.rsapi.adapter.web.InboxResponseMapper.toHttpStatus
import org.railwaystations.rsapi.adapter.web.RequestUtil
import org.railwaystations.rsapi.adapter.web.model.AdminInboxCommandResponseDto
import org.railwaystations.rsapi.adapter.web.model.InboxCommandDto
import org.railwaystations.rsapi.adapter.web.model.InboxCommandDto.Command
import org.railwaystations.rsapi.adapter.web.model.InboxCommandDto.ConflictResolution
import org.railwaystations.rsapi.adapter.web.model.InboxCountResponseDto
import org.railwaystations.rsapi.adapter.web.model.InboxEntryDto
import org.railwaystations.rsapi.adapter.web.model.InboxResponseDto
import org.railwaystations.rsapi.adapter.web.model.InboxStateQueryRequestDto
import org.railwaystations.rsapi.adapter.web.model.InboxStateQueryResponseDto
import org.railwaystations.rsapi.adapter.web.model.InboxStateQueryResponseDto.State
import org.railwaystations.rsapi.adapter.web.model.ProblemReportDto
import org.railwaystations.rsapi.adapter.web.model.ProblemReportTypeDto
import org.railwaystations.rsapi.adapter.web.model.PublicInboxEntryDto
import org.railwaystations.rsapi.core.model.Coordinates
import org.railwaystations.rsapi.core.model.InboxCommand
import org.railwaystations.rsapi.core.model.InboxEntry
import org.railwaystations.rsapi.core.model.InboxStateQuery
import org.railwaystations.rsapi.core.model.InboxStateQuery.InboxState
import org.railwaystations.rsapi.core.model.ProblemReport
import org.railwaystations.rsapi.core.model.ProblemReportType
import org.railwaystations.rsapi.core.model.PublicInboxEntry
import org.railwaystations.rsapi.core.ports.ManageInboxUseCase
import org.railwaystations.rsapi.core.ports.ManageProfileUseCase
import org.railwaystations.rsapi.utils.Logger
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestHeader
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.servlet.LocaleResolver
import java.util.*

@RestController
@Validated
class InboxController(
    private val manageInboxUseCase: ManageInboxUseCase,
    private val manageProfileUseCase: ManageProfileUseCase,
    private val localeResolver: LocaleResolver,
    private val requestUtil: RequestUtil,
) {

    private val log by Logger()

    private fun toDomain(problemReport: ProblemReportDto): ProblemReport {
        return ProblemReport(
            countryCode = problemReport.countryCode,
            stationId = problemReport.stationId,
            title = problemReport.title,
            photoId = problemReport.photoId,
            type = toDomain(problemReport.type),
            comment = problemReport.comment,
            coordinates = mapCoordinates(problemReport.lat, problemReport.lon),
        )
    }

    private fun toDomain(dtoType: ProblemReportTypeDto): ProblemReportType {
        return when (dtoType) {
            ProblemReportTypeDto.OTHER -> ProblemReportType.OTHER
            ProblemReportTypeDto.WRONG_NAME -> ProblemReportType.WRONG_NAME
            ProblemReportTypeDto.WRONG_PHOTO -> ProblemReportType.WRONG_PHOTO
            ProblemReportTypeDto.PHOTO_OUTDATED -> ProblemReportType.PHOTO_OUTDATED
            ProblemReportTypeDto.STATION_ACTIVE -> ProblemReportType.STATION_ACTIVE
            ProblemReportTypeDto.STATION_INACTIVE -> ProblemReportType.STATION_INACTIVE
            ProblemReportTypeDto.WRONG_LOCATION -> ProblemReportType.WRONG_LOCATION
            ProblemReportTypeDto.STATION_NONEXISTENT -> ProblemReportType.STATION_NONEXISTENT
            ProblemReportTypeDto.DUPLICATE -> ProblemReportType.DUPLICATE
        }
    }

    private fun mapCoordinates(lat: Double?, lon: Double?): Coordinates? {
        if (lat == null || lon == null) {
            return null
        }
        return Coordinates(lat, lon)
    }

    private fun toPublicInboxEntryDto(publicInboxEntries: List<PublicInboxEntry>): List<PublicInboxEntryDto> {
        return publicInboxEntries.map { publicInboxEntry -> toDto(publicInboxEntry) }
    }

    private fun toDto(publicInboxEntry: PublicInboxEntry): PublicInboxEntryDto {
        return PublicInboxEntryDto(
            countryCode = publicInboxEntry.countryCode,
            stationId = publicInboxEntry.stationId,
            title = publicInboxEntry.title,
            lat = publicInboxEntry.lat,
            lon = publicInboxEntry.lon
        )
    }

    private fun toIdList(inboxStateQueryRequestDtos: List<InboxStateQueryRequestDto>): List<Long> {
        return inboxStateQueryRequestDtos.map { it.id }
    }

    private fun toInboxStateQueryDto(inboxStateQueries: List<InboxStateQuery>): List<InboxStateQueryResponseDto> {
        return inboxStateQueries.map { toDto(it) }
    }

    private fun toDto(inboxStateQuery: InboxStateQuery): InboxStateQueryResponseDto {
        return InboxStateQueryResponseDto(
            id = inboxStateQuery.id,
            state = toDto(inboxStateQuery.state),
            countryCode = inboxStateQuery.countryCode,
            stationId = inboxStateQuery.stationId,
            title = inboxStateQuery.title,
            lat = inboxStateQuery.coordinates?.lat,
            lon = inboxStateQuery.coordinates?.lon,
            newTitle = inboxStateQuery.newTitle,
            newLat = inboxStateQuery.newCoordinates?.lat,
            newLon = inboxStateQuery.newCoordinates?.lon,
            comment = inboxStateQuery.comment,
            problemReportType = toDto(inboxStateQuery.problemReportType),
            rejectedReason = inboxStateQuery.rejectedReason,
            filename = inboxStateQuery.filename,
            inboxUrl = inboxStateQuery.inboxUrl,
            crc32 = inboxStateQuery.crc32,
            createdAt = inboxStateQuery.createdAt?.toEpochMilli(),
        )
    }

    private fun toDto(inboxState: InboxState): State {
        return when (inboxState) {
            InboxState.REVIEW, InboxState.CONFLICT -> State.REVIEW
            InboxState.ACCEPTED -> State.ACCEPTED
            InboxState.REJECTED -> State.REJECTED
            InboxState.UNKNOWN -> State.UNKNOWN
        }
    }

    private fun toInboxEntryDto(inboxEntries: List<InboxEntry>): List<InboxEntryDto> {
        return inboxEntries.map { toDto(it) }
    }

    private fun toDto(inboxEntry: InboxEntry): InboxEntryDto {
        return InboxEntryDto(
            id = inboxEntry.id,
            photographerNickname = inboxEntry.photographerNickname!!,
            comment = inboxEntry.comment ?: "",
            createdAt = inboxEntry.createdAt!!.toEpochMilli(),
            done = inboxEntry.done,
            hasPhoto = inboxEntry.hasPhoto(),
            countryCode = inboxEntry.countryCode,
            stationId = inboxEntry.stationId,
            title = inboxEntry.title,
            lat = inboxEntry.lat,
            lon = inboxEntry.lon,
            newTitle = inboxEntry.newTitle,
            newLat = inboxEntry.newLat,
            newLon = inboxEntry.newLon,
            photographerEmail = inboxEntry.photographerEmail,
            photoId = inboxEntry.photoId,
            filename = inboxEntry.filename,
            inboxUrl = inboxEntry.inboxUrl,
            hasConflict = inboxEntry.conflict,
            problemReportType = toDto(inboxEntry.problemReportType),
            isProcessed = inboxEntry.processed,
            active = inboxEntry.active,
        )
    }

    private fun toDto(problemReportType: ProblemReportType?): ProblemReportTypeDto? {
        return when (problemReportType) {
            ProblemReportType.STATION_NONEXISTENT -> ProblemReportTypeDto.STATION_NONEXISTENT
            ProblemReportType.WRONG_LOCATION -> ProblemReportTypeDto.WRONG_LOCATION
            ProblemReportType.STATION_ACTIVE -> ProblemReportTypeDto.STATION_ACTIVE
            ProblemReportType.STATION_INACTIVE -> ProblemReportTypeDto.STATION_INACTIVE
            ProblemReportType.PHOTO_OUTDATED -> ProblemReportTypeDto.PHOTO_OUTDATED
            ProblemReportType.WRONG_PHOTO -> ProblemReportTypeDto.WRONG_PHOTO
            ProblemReportType.WRONG_NAME -> ProblemReportTypeDto.WRONG_NAME
            ProblemReportType.OTHER -> ProblemReportTypeDto.OTHER
            ProblemReportType.DUPLICATE -> ProblemReportTypeDto.DUPLICATE
            null -> null
        }
    }

    private fun toDomain(command: InboxCommandDto): InboxCommand {
        return InboxCommand(
            id = command.id,
            countryCode = command.countryCode,
            stationId = command.stationId,
            title = command.title,
            coordinates = mapCoordinates(command.lat, command.lon),
            rejectReason = command.rejectReason,
            ds100 = command.DS100,
            active = command.active,
            conflictResolution = toDomain(command.conflictResolution),
        )
    }

    private fun toDomain(conflictResolution: ConflictResolution?): InboxCommand.ConflictResolution {
        return when (conflictResolution) {
            null -> InboxCommand.ConflictResolution.DO_NOTHING
            ConflictResolution.DO_NOTHING -> InboxCommand.ConflictResolution.DO_NOTHING
            ConflictResolution.OVERWRITE_EXISTING_PHOTO -> InboxCommand.ConflictResolution.OVERWRITE_EXISTING_PHOTO
            ConflictResolution.IMPORT_AS_NEW_PRIMARY_PHOTO -> InboxCommand.ConflictResolution.IMPORT_AS_NEW_PRIMARY_PHOTO
            ConflictResolution.IMPORT_AS_NEW_SECONDARY_PHOTO -> InboxCommand.ConflictResolution.IMPORT_AS_NEW_SECONDARY_PHOTO
            ConflictResolution.IGNORE_NEARBY_STATION -> InboxCommand.ConflictResolution.IGNORE_NEARBY_STATION
        }
    }

    @PreAuthorize("hasRole('ADMIN')")
    @RequestMapping(
        method = [RequestMethod.GET],
        value = ["/adminInboxCount"],
        produces = ["application/json"]
    )
    fun adminInboxCountGet(): ResponseEntity<InboxCountResponseDto> {
        return ResponseEntity.ok(InboxCountResponseDto(manageInboxUseCase.countPendingInboxEntries()))
    }

    @PreAuthorize("hasRole('ADMIN')")
    @RequestMapping(
        method = [RequestMethod.GET],
        value = ["/adminInbox"],
        produces = ["application/json"]
    )
    fun adminInboxGet(
        @RequestHeader(
            required = true,
            value = "Authorization"
        ) authorization: String
    ): ResponseEntity<List<InboxEntryDto>> {
        return ResponseEntity.ok(toInboxEntryDto(manageInboxUseCase.listAdminInbox(requestUtil.authUser.user)))
    }

    @PreAuthorize("hasRole('ADMIN')")
    @RequestMapping(
        method = [RequestMethod.POST],
        value = ["/adminInbox"],
        produces = ["application/json"],
        consumes = ["application/json"]
    )
    fun adminInboxPost(
        @RequestHeader(required = true, value = "Authorization") authorization: String,
        @Valid @RequestBody inboxCommandDto: InboxCommandDto
    ): ResponseEntity<AdminInboxCommandResponseDto> {
        log.info(
            "Executing adminInbox commandDto {} for Nickname: {}",
            inboxCommandDto.command,
            requestUtil.authUser.username
        )
        try {
            val command: InboxCommand = toDomain(inboxCommandDto)
            when (inboxCommandDto.command) {
                Command.REJECT -> manageInboxUseCase.rejectInboxEntry(command)
                Command.IMPORT_PHOTO -> manageInboxUseCase.importPhoto(command)
                Command.IMPORT_MISSING_STATION -> manageInboxUseCase.importMissingStation(command)
                Command.ACTIVATE_STATION -> manageInboxUseCase.updateStationActiveState(
                    command,
                    true
                )

                Command.DEACTIVATE_STATION -> manageInboxUseCase.updateStationActiveState(
                    command,
                    false
                )

                Command.DELETE_STATION -> manageInboxUseCase.deleteStation(command)
                Command.DELETE_PHOTO -> manageInboxUseCase.deletePhoto(command)
                Command.MARK_SOLVED -> manageInboxUseCase.markProblemReportSolved(command)
                Command.CHANGE_NAME -> manageInboxUseCase.changeStationTitle(command)
                Command.UPDATE_LOCATION -> manageInboxUseCase.updateLocation(command)
                Command.PHOTO_OUTDATED -> manageInboxUseCase.markPhotoOutdated(command)
            }
        } catch (e: IllegalArgumentException) {
            log.warn("adminInbox commandDto {} failed", inboxCommandDto, e)
            return ResponseEntity<AdminInboxCommandResponseDto>(
                AdminInboxCommandResponseDto(
                    HttpStatus.BAD_REQUEST.value(),
                    e.message ?: e.toString()
                ), HttpStatus.BAD_REQUEST
            )
        }

        return ResponseEntity<AdminInboxCommandResponseDto>(
            AdminInboxCommandResponseDto(HttpStatus.OK.value(), "ok"),
            HttpStatus.OK
        )
    }

    @RequestMapping(
        method = [RequestMethod.GET],
        value = ["/publicInbox"],
        produces = ["application/json"]
    )
    fun publicInboxGet(): ResponseEntity<List<PublicInboxEntryDto>> {
        return ResponseEntity.ok(toPublicInboxEntryDto(manageInboxUseCase.publicInbox()))
    }

    @PreAuthorize("isAuthenticated()")
    @RequestMapping(
        method = [RequestMethod.POST],
        value = ["/reportProblem"],
        produces = ["application/json"]
    )
    fun reportProblemPost(
        @RequestHeader(required = true, value = "Authorization") authorization: String,
        @Valid @RequestBody problemReport: ProblemReportDto
    ): ResponseEntity<InboxResponseDto> {
        val locale: Locale = localeResolver.resolveLocale(requestUtil.request)
        val user = requestUtil.authUser.user
        if (user.locale != locale) {
            manageProfileUseCase.updateLocale(user, locale)
        }

        val inboxResponse: InboxResponseDto =
            InboxResponseMapper.toDto(
                manageInboxUseCase.reportProblem(
                    toDomain(problemReport),
                    user,
                    requestUtil.userAgent
                )
            )
        return ResponseEntity<InboxResponseDto>(inboxResponse, toHttpStatus(inboxResponse.state))
    }

    @PreAuthorize("isAuthenticated()")
    @RequestMapping(
        method = [RequestMethod.GET],
        value = ["/userInbox"],
        produces = ["application/json"]
    )
    fun userInboxGet(
        @RequestHeader(
            required = true,
            value = "Authorization"
        ) authorization: String
    ): ResponseEntity<List<InboxStateQueryResponseDto>> {
        return ResponseEntity.ok(
            toInboxStateQueryDto(
                manageInboxUseCase.userInbox(
                    requestUtil.authUser.user
                )
            )
        )
    }

    @PreAuthorize("isAuthenticated()")
    @RequestMapping(
        method = [RequestMethod.POST],
        value = ["/userInbox"],
        produces = ["application/json"],
        consumes = ["application/json"]
    )
    fun userInboxPost(
        @RequestHeader(required = true, value = "Authorization") authorization: String,
        @Valid @RequestBody uploadStateQueries: List<InboxStateQueryRequestDto>
    ): ResponseEntity<List<InboxStateQueryResponseDto>> {
        return ResponseEntity.ok(
            toInboxStateQueryDto(
                manageInboxUseCase.userInbox(
                    requestUtil.authUser.user,
                    toIdList(uploadStateQueries)
                )
            )
        )
    }
}
